src/bcounter_mgr.erl

```diff
@@ -31,201 +31,294 @@
+-type antidote_crdt_counter_b_downstream_op_result_or_error() :: antidote_crdt_counter_b_downstream_op_result() | {error, no_permission}. %%TODO maybe too specific here
 
 %% ===================================================================
 %% Public API
 %% ===================================================================
 
-start_link() ->
-    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
-
-init([]) ->
-    Timer=erlang:send_after(?TRANSFER_FREQ, self(), transfer_periodic),
-    {ok, #state{req_queue=orddict:new(), transfer_timer=Timer, last_transfers=orddict:new()}}.
-
 %% @doc Processes a decrement operation for a bounded counter.
 %% If the operation is unsafe (i.e. the value of the counter can go
-%% below 0), operation fails, otherwhise a downstream for the decrement
+%% below 0), operation fails, otherwise a downstream for the decrement
 %% is generated.
-generate_downstream(Key, {decrement, {V, _}}, BCounter) ->
-    MyDCId = dc_utilities:get_my_dc_id(),
-    gen_server:call(?MODULE, {consume, Key, {decrement, {V, MyDCId}}, BCounter});
+-spec generate_downstream({key(), bucket()}, antidote_crdt_counter_b:antidote_crdt_counter_b_anon_op(), antidote_crdt_counter_b:antidote_crdt_counter_b()) -> antidote_crdt_counter_b_downstream_op_result_or_error() | {error, invalid_dcid}.
+generate_downstream(Key, {decrement, {Amount, _IgnoreDCID}}, BCounter) ->
+    MyDCID = dc_utilities:get_my_dc_id(),
+    gen_server:call(?MODULE, {Key, {decrement, {Amount, MyDCID}}, BCounter});
 
 %% @doc Processes an increment operation for the bounded counter.
 %% Operation is always safe.
-generate_downstream(_Key, {increment, {Amount, _}}, BCounter) ->
-    MyDCId = dc_utilities:get_my_dc_id(),
-    ?DATA_TYPE:downstream({increment, {Amount, MyDCId}}, BCounter);
+generate_downstream(_Key, {increment, {Amount, _IgnoreDCID}}, BCounter) ->
+    MyDCID = dc_utilities:get_my_dc_id(),
+    antidote_crdt_counter_b:downstream({increment, {Amount, MyDCID}}, BCounter);
 
-%% @doc Processes a trasfer operation between two owners of the
+%% @doc Processes a transfer operation between two owners of the
 %% counter.
-generate_downstream(_Key, {transfer, {Amount, To, From}}, BCounter) ->
-    ?DATA_TYPE:downstream({transfer, {Amount, To, From}}, BCounter).
+generate_downstream(_Key, {transfer, {Amount, ToDCID, _IgnoreFromDCID}}, BCounter) ->
+    case check_valid_dcid(ToDCID) of %%prevent transferring to arbitrary DCIDs
+        true ->
+            MyDCID = dc_utilities:get_my_dc_id(),
+            case ToDCID /= MyDCID of
+                true -> antidote_crdt_counter_b:downstream({transfer, {Amount, ToDCID, MyDCID}}, BCounter);
+                false ->
+                    %%TODO this transfer is equal to increment
+                    generate_downstream(_Key, {increment, {Amount, MyDCID}}, BCounter)
+            end;
+        false -> {error, invalid_dcid}
+    end.
```

Comment:
Here I added some restrictions
-> Transfer only allowed from MyDCID
-> Transfer only allowed to valid DCIDs (descriptor exists)
-> Self-transfer is still allowed but it is made explicit in the code that this is equal to an increment and changes the total permissions of the bcounter